Extended duration symbols in ABC chords (marian-schultz)

marian-schulz opened the pull request:
Extension for the recognition of chords in ABCFormat according to http://abcnotation.com/wiki/abc:standard:v2.1#chords_and_unisons.

* For chords an outer length modifier is supported
Example:
[ceg]2 == [c2e2g2]
[c3e3g3]2 == [ceg]6 == [c6e6g6]

* The length of the chord is determined by its first note
Example:
[c2e4g8] == [c2e2g2]
* Empty chords are discarded
* Additional unittests for variations of abc Chords
* Fix: Missing chord symbols  for chords
* Additional unittests for chord symbols on note & chords

mscuthbert commented:
coveralls had a problem -- closing and reopening to get a fresh test.

mscuthbert commented:
Thanks -- look at the lint errors, etc. -- but it seems to be on the right track!

marian-schulz commented:
I have fixed the PEP8 violations and cleaned up my commit history.
The tests are all green now.

I apologize for not doing the code inspection before.

marian-schulz commented:
The chord symbols before a chord were missing. I have also fixed this and added an additional unittest.

mscuthbert commented on music21/abcFormat/testFiles.py:
```diff
@@ -816,6 +816,67 @@ def testAbc21DirectiveCarryNot(self):
+        for abc_chord in ['[]', '[z]']:
+            ah = af.readstr(abc_dl + '[]')
+            s = translate.abcToStreamScore(ah)
+            self.assertFalse(s.elements[1].getElementsByClass(chord.Chord),
```
see below about elements[1]

mscuthbert commented on music21/abcFormat/testFiles.py:
```diff
@@ -816,6 +816,67 @@ def testAbc21DirectiveCarryNot(self):
+            s = translate.abcToStreamScore(ah)
+            self.assertEqual(s.duration.quarterLength, quarter_length,
+                             'invalid duration of chord "%s"' % abc_chord)
+            for e in s.elements[1].getElementsByClass(chord.Chord):
```
can `.elements[1]` be changed to whatever it is we're actually getting, like `.parts[0]` etc. -- otherwise this test is fragile if other metadata elements or something else like a spanner is later a default part of abcToStreamScore.

marian-schulz replied:
Fixed this in all cases. I use the getElementbyClass of operator now.

mscuthbert commented on music21/abcFormat/testFiles.py:
```diff
@@ -816,6 +816,67 @@ def testAbc21DirectiveCarryNot(self):
+            ('[ce2g]', 0.5),
+            ('[ceg2]', 0.5),
+            ('[c2e2g2]/2', 0.5),
+            ('[c/2e/2g/2]', 0.25),
```
Add some tests with sharps, accidentals, etc. to make sure that c#/2 still works.  Maybe a separate test so it doesn't break the pitch_name in 'CEG' test below.  Just always paranoid about that.

marian-schulz replied:
I agree.

mscuthbert commented on music21/abcFormat/testFiles.py:
```diff
@@ -816,6 +816,67 @@ def testAbc21DirectiveCarryNot(self):
+                    self.assertIn(pitch_name, e.pitchNames,
+                                  'Pitch not in Chord "%s"' % abc_chord)
+
+    def testAbc21ChordSymbol(self):
```
great!

mscuthbert reviewed (changes requested):
Looks great!  The `.elements[1]` call is all that needs to be changed, to make sure that it's explicitly getting the type of element that it should be.

(btw -- `s.elements[1]` is almost always better as `e[1]` anyhow -- .elements[1] is slower and doesn't setActiveSite).

marian-schulz commented:
Optimze the Testcases.
+ Use the ElementsByClass operator instead of an index access to .elements[x].
+ Add sharps, flats, natural and ocatve modificators to notes.

Broken rythm
+ Fixes broken rythm for Chords (i guess i broke that)
+ Add A Testcase to cover brocken rythm between chords and notes

jacobtylerwalls commented on music21/abcFormat/__init__.py:
```diff
@@ -1641,21 +1641,37 @@ class ABCChord(ABCNote):
+            Chord without length modifier: [ceg]
+            Chords with outer length modifier: [ceg]2, [ceg]/2
+            Cords with inner length modifier: [c2e2g2], [c2eg]
+            Cords with inner and outer length modifier: [c2e2g2]/2, [c/2e/2g/2]2
```
Chords

jacobtylerwalls commented on music21/abcFormat/testFiles.py:
```diff
@@ -816,10 +816,127 @@ def testAbc21DirectiveCarryNot(self):
 if __name__ == '__main__':
     import music21
-    # music21.converter.parse(reelsABC21, format='abc').scores[1].show()
+    music21.converter.parse(reelsABC21, format='abc').scores[1].show()
```
Better to leave this commented so it doesn't launch a file.

jacobtylerwalls reviewed (commented):
Happened to stroll by and noticed a couple small typos or details.

mscuthbert commented on music21/abcFormat/testFiles.py:
```diff
@@ -831,7 +831,8 @@ def testAbc21Chords(self):
             ah = af.readstr(abc_dl + '[]')
             s = translate.abcToStreamScore(ah)
-            self.assertFalse(s.elements[1].getElementsByClass(chord.Chord),
+            part = s.getElementsByClass(stream.Part)
```
this is returning a Stream iterator -- you probably want a specific part, like:

```
part = s.getElementsByClass(stream.Part)[0]
```

which is abbreviated as:

```
part = s.parts[0]
```

marian-schulz replied:
Thanks,
I fixed this and  notice an  additional Bug, i guess made by the translater. (I still have to research this)

for abc_chord in ['[]', '[z]']:
    ah = af.readstr(abc_dl + '[]')
    s = translate.abcToStreamScore(ah)
    chord = list(s.parts[0].notes)
    print(abc_chord, chord, ah.tokens)

[] [<music21.note.Note C>] [<music21.abcFormat.ABCMetadata 'L:1/8'>, <music21.abcFormat.ABCChord '[]'>]
[z] [<music21.note.Note C>] [<music21.abcFormat.ABCMetadata 'L:1/8'>, <music21.abcFormat.ABCChord '[]'>]

There is no Chord at all but a note C in the part. The C is not present in abc code nor token list at all.

mscuthbert commented on music21/abcFormat/testFiles.py:
```diff
@@ -854,14 +855,15 @@ def testAbc21Chords(self):
                              'invalid duration of chord "%s"' % abc_chord)
-            for e in s.elements[1].getElementsByClass(chord.Chord):
+            part = s.getElementsByClass(stream.Part)
+            for e in part.getElementsByClass(chord.Chord):
```
Similarly here.  I'll bet that `e` is never being set.  Temporarily put a print statement into the loop and you'll see it never runs.

marian-schulz replied:
I wonder while the stream iterator implements .getElementsByClass at all ?
However .parts[0] looks better.
fixed in the last commit.

mscuthbert replied:
It totally does.  StreamIterators are designed to be chained, so most explicit cases would be:

```
s.flat.iter.getElementsByOffset(10.0, 20.0).getElementsByClass('Note').getElementsByGroup('cue').addFilter(lambda n: n.pitch.alter == 1)
```

As of v.6., each of these "getElementsBy..." returns a new StreamIterator that adds a new filter on the notes.  So `s.getElementsByClass(stream.Part).getElementsByClass(chord.Chord)` would always return no elements, because first it has to be of class Part, and then of class Chord, and nothing will ever be like that.

It might be that adding a second filter of the same type should raise an exception, but there are a few cases where you might want this, such as:

`s.getElementsByClass(spanner.Spanner).getElementsByClass(expressions.Tremolo)`

which only gets Tremolos that are Spanners. 

Maybe that doesn't work, but there are a few cases like that.

mscuthbert commented on music21/abcFormat/__init__.py:
```diff
@@ -1669,9 +1669,9 @@ def parse(self, forceKeySignature=None, forceDefaultQuarterLength=None):
-                                                         forceDefaultQuarterLength=1.0)
+
+        outer_lengthModifier = self.getQuarterLength(outerLengthModifierStr,
+                                                    forceDefaultQuarterLength=1.0)
```
misalignment of hanging indent (one space more)

marian-schulz replied:
add this extra space and removed the previous line.

mscuthbert commented:
looks really good -- one little change -- capital letter variable names are only used for classes.  Will merge if everything passes.  Thank you for the great contributions!

jacobtylerwalls commented on music21/abcFormat/testFiles.py:
```diff
@@ -816,10 +816,130 @@ def testAbc21DirectiveCarryNot(self):
+                self.assertIn(pitch_name, chord_symbol[0].pitchNames,
+                              'Pitch not in ChordSymbol of abc: "%s"' % abc_text)
+
+    def testAbc21BrokenRythm(self):
```
rhythm

jacobtylerwalls commented on music21/abcFormat/testFiles.py:
```diff
@@ -816,10 +816,130 @@ def testAbc21DirectiveCarryNot(self):
+        # default length of this test
+        abc_dl = 'L:1/4\n'
+
+        # test abc strings of broken rythm between 2 notes and/or chords and their
```
same

jacobtylerwalls commented on music21/abcFormat/testFiles.py:
```diff
@@ -816,10 +816,130 @@ def testAbc21DirectiveCarryNot(self):
+            self.assertEqual(ist_left.duration.quarterLength, soll_left,
+                             f'Invalid left note/chord length of abc broken rythm: {abc}')
+            self.assertEqual(ist_right.duration.quarterLength, soll_right,
+                             f'Invalid right note/chord length of abc broken rythm: {abc}')
```
here and above

jacobtylerwalls reviewed (commented):
Noticed a few instances where "rhythm" is misspelled.

mscuthbert commented:
Please don't force-push when making Pull Requests -- it makes it impossible to review what has changed.  Your open-source maintainers will appreciate it.

mscuthbert commented:
Congrats! Merged
